modules/idlgen/core/src/main/scala/model.scala

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017-2018 47 Degrees, LLC. <http://www.47deg.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.rpc.idlgen.protobuf
+
+import freestyle.rpc.protocol._
+
+object model {
+
+  sealed trait ProtoFieldMod extends Product with Serializable
+
+  case object Repeated extends ProtoFieldMod
+
+  sealed trait ProtoMessageField extends Product with Serializable {
+    def mod: Option[ProtoFieldMod]
+
+    def name: String
+
+    def id: String
+
+    def tag: Int
+
+    def comment: Option[String]
+  }
+
+  final case class ProtoDouble(
```

Comment:
In this `ProtoMessageField` abstract-data-type, each and every one of the classes below overrides the `def` in the trait with a default parameter `id`. However, ¿is there a reason why there should be an instance of, say, `ProtoFloat` with a value of `id` different to `"float"`?

If that value would always be the same, it could make more sense to use something else than a constructor parameter. For instance, it could be an `override val` inside the body of the class. Or, you could make `ProtoMessageField` into an abstract class with one constructor parameter `id` (or with all fields as constructor parameters).

Reply:
Thanks for reviewing! Agreed that the current code isn't optimal - I left it as-is when migrating from the plugin project so we could focus on the migration/macro-deduplication parts. I have a code refactoring PR ready to push once this is merged, that will address this by using a model that's less tied to Protobufs and will better support Avro once implemented.